src/main.cpp

```diff
@@ -4,41 +4,152 @@
 #include "lua/src/lua.hpp"
 #include "peglib.h"
 
+#define DEBUG_PARSER
 
 using namespace peg;
 using namespace std;
 
+// utilities for lua stack manipulation
+struct LuaStackPtr{
+	string varName;
+	int stackIndex=0;
+};
+
+void lua_push(lua_State *L, const int value) {lua_pushinteger(L, value);}
+void lua_push(lua_State *L, const string value) {lua_pushstring(L, value.c_str());}
+void lua_push(lua_State *L, const char *value) {lua_pushstring(L, value);}
+void lua_push(lua_State *L, const char *value, size_t len) {lua_pushlstring(L, value, len);}
+void lua_closefield(lua_State *L){ lua_settable(L, -3); }
+
+template<typename TKey, typename TVal>
+void lua_pushfield(lua_State *L, TKey key, TVal value){
+	lua_push(L, key);
+	lua_push(L, value);
+	lua_closefield(L);
+}
+
+template<typename TKey>
+void lua_pushfield(lua_State *L, TKey key, const char *value, size_t len){
+	lua_push(L, key);
+	lua_pushlstring(L, value, len);
+	lua_closefield(L);
+}
 
+void lua_pushfield(lua_State *L, LuaStackPtr ptr){
+	lua_push(L, ptr.varName);
+	lua_pushvalue(L, ptr.stackIndex);
+	lua_closefield(L);
+}
+
+template<typename TKey>
+void lua_opensubtable(lua_State *L, TKey key){
+	lua_push(L, key);
+	lua_newtable(L);
+}
+
+// main
 int Main(vector<string> args)
 {
 
-	// parser test:
-	//-------------
-	
-	// create grammar
-	const auto syntax = R"(
-		Additive <- (Multitive AddOp)* Multitive
-		AddOp <- '+' / '-'
-		Multitive <- (Primary MulOp)* Primary
-		MulOp <- '*' / '/'
-		Primary <- '(' Additive ')'
-			 / Number
-		Number <- < [0-9]+ >
-		%whitespace <- [ \t]*
-	)";
+	// init lua
+	lua_State *L = luaL_newstate();
+	luaL_openlibs(L);
 
-	// create parser
-	parser parser(syntax);
+	// load parser script
+	auto result = luaL_loadfile(L, "parser.lua") || lua_pcall(L, 0, 0, 0);
+	if (result){
+		cout << "error loading parser.lua: %s!\n", lua_tostring(L, -1);
+		return EXIT_FAILURE;
+	}
 
-	if (!parser){
-		cout << "error creating grammar" << endl;
+	// copy grammar from parser script
+	lua_getglobal(L, "grammar");
```

Comment:
I think it would be even nicer if this module exposed a `parser` class to the Lua code. The Lua code could then create a parser by passing a grammar (string), a table of callbacks and a default callback to the constructor. The returned object would expose `parse` methods in order to parse strings, streams or whatever other appropriate abstraction Lua provides.